### PR TITLE
kube-proxy react on Node PodCIDR changes

### PR DESCRIPTION
**Pull request by code-elinka**
Cherry pick of #111344 on release-1.25.

#111344: kube-proxy: gate topology correctly

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
kube-proxy, will restart in case it detects that the Node assigned pod.Spec.PodCIDRs have changed
```

### CODE REVIEW

**Comment by code-elinka** [comment]
@aojea I added a line with `"k8s.io/klog/v2"` because it failed without it. I am also checking the development setup to test it locally

**Comment by aojea** [comment]
> @aojea I added a line with `"k8s.io/klog/v2"` because it failed without it. I am also checking the development setup to test it locally

it's ok, it is embedded with the other imports, it seems it was moved out in one of the refactors, no worries

**Comment by aojea** [comment]
unrelated failures and known flaky test

> Kubernetes e2e suite: [It] [sig-node] Probing container should mark readiness on pods to false while pod is in progress of terminating when a pod has a readiness probe expand_less

**Comment by aojea** [comment]
/retitle kube-proxy react on Node PodCIDR changes

File: pkg/proxy/topology.go [hunk 1]
```diff
@@ -145,6 +145,9 @@ func CategorizeEndpoints(endpoints []Endpoint, svcInfo ServicePort, nodeLabels m
 // * All of the endpoints for this Service have a topology hint
 // * At least one endpoint for this Service is hinted for this node's zone.
 func canUseTopology(endpoints []Endpoint, svcInfo ServicePort, nodeLabels map[string]string) bool {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.TopologyAwareHints) {
```
Comment (khenidak) [inline pkg/proxy/topology.go:148]: I would separate this in a different PR. Once we do we can approve and move this forward.

Comment (aojea) [reply pkg/proxy/topology.go:148]: what is the difference, they are in different commits already?

**Comment by aojea** [comment]
ping @kubernetes/release-managers

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve
